packages/react-reconciler/src/ReactFiberThrow.new.js

```diff
@@ -297,76 +297,82 @@ function throwException(
           wakeables.add(wakeable);
         }
 
-        // If the boundary is in legacy mode, we should *not*
-        // suspend the commit. Pretend as if the suspended component rendered
-        // null and keep rendering. In the commit phase, we'll schedule a
-        // subsequent synchronous update to re-render the Suspense.
-        //
-        // Note: It doesn't matter whether the component that suspended was
-        // inside a concurrent mode tree. If the Suspense is outside of it, we
-        // should *not* suspend the commit.
-        //
-        // If the suspense boundary suspended itself suspended, we don't have to
-        // do this trick because nothing was partially started. We can just
-        // directly do a second pass over the fallback in this render and
-        // pretend we meant to render that directly.
-        if (
-          (workInProgress.mode & ConcurrentMode) === NoMode &&
-          workInProgress !== returnFiber
-        ) {
-          workInProgress.flags |= DidCapture;
-          sourceFiber.flags |= ForceUpdateForLegacySuspense;
+        if ((workInProgress.mode & ConcurrentMode) === NoMode) {
+          // Legacy Mode Suspense
+          //
+          // If the boundary is in legacy mode, we should *not*
+          // suspend the commit. Pretend as if the suspended component rendered
+          // null and keep rendering. When the Suspense boundary completes,
+          // we'll do a second pass to render the fallback.
+          if (workInProgress === returnFiber) {
+            // Special case where we suspended while reconciling the children of
+            // a Suspense boundary's inner Offscreen wrapper fiber. This happens
+            // when a React.lazy component is a direct child of a
+            // Suspense boundary.
+            //
+            // Suspense boundaries are implemented as multiple fibers, but they
+            // are a single conceptual unit. The legacy mode behavior where we
+            // pretend the suspended fiber committed as `null` won't work,
+            // because in this case the "suspended" fiber is the inner
+            // Offscreen wrapper.
+            //
+            // Because the contents of the boundary haven't started rendering
+            // yet (i.e. nothing in the tree has partially rendered) we can
+            // switch to the regular, concurrent mode behavior: mark the
+            // boundary with ShouldCapture and enter the unwind phase.
+            workInProgress.flags |= ShouldCapture;
```

Comment:
The profiler call is a red herring. The relevant part of this branch is that it marks a `DidCapture` flag:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberUnwindWork.new.js#L117

and it returns the work-in-progress fiber (which in this case is a Suspense boundary) instead of returning `null`:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberUnwindWork.new.js#L125-L127

That will tell the work loop to render the Suspense boundary again:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberWorkLoop.new.js#L1700-L1708

And since we set a `DidCapture` flag, we'll render the fallback this time instead of the regular children:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberBeginWork.new.js#L1936

Reply:
If you look at the old code, though, we were already doing this because line 315 here causes us to fall through to the regular concurrent mode path:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberThrow.new.js#L313-L316

which looks like this:

https://github.com/facebook/react/blob/d174d063d13d428ad4002474cb560464a7f70f45/packages/react-reconciler/src/ReactFiberThrow.new.js#L412-L414

So the net change for the branch I added is:

```diff
- attachPingListener(root, wakeable, rootRenderLanes);
workInProgress.flags |= ShouldCapture;
```

Reply:
Ahh okay makes sense!